### PR TITLE
[vmap] fix reduction boxed batching rules

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91109

Fixes https://github.com/pytorch/pytorch/issues/91041

There's a bug in our boxed reduction batching rules for a very specific
case: vmap over a Tensor of shape [1] for an operation where the
output rank is supposed to be less than the input rank, e.g.

```
x = torch.tensor([10.], device=device)
y = vmap(lambda x: x.sum(0))(x)
```

The boxed reduction batching rule handles three types of "reduction"
operations:
- reduction operations with an optional keepdim argument, which
specifies if the output should have the same or smaller rank than the
input
- reduction operations without a keepdim arg that morally have keepdim=True (like cumsum --
which never actually modifies the rank of the tensor but is still a
"reduction" since it sums a bunch of things together)
- reduction operations without a keepdim arg that morally have
keepdim=False. (just torch.count_nonzero).

Furthermore, PyTorch has special handling for scalar tensors (e.g.
tensors of shape []). It is valid to do
`torch.sum(torch.tensor(10.), dim=0)`.

This PR updates the `boxed_reduction_batch_rule` to handle the
interaction between the three kinds of reduction and the scalar tensor
cases correctly. Concretely, it:
- introduces additional templates to `boxed_reduction_batch_rule` for
what type of "keepdim" reduction this is.
- splits the old REDUCTION_BOXED macro (which was a good default) into
REDUCTION_NO_KEEPDIM_ARG and REDUCTION_WITH_KEEPDIM_ARG (which are also
opionated defaults) and uses them.

Test Plan:
- Given an input of shape [], our vmap OpInfo test suite only produces
a Tensor of shape [B] with B = 2. At first glance this doesn't look
sufficient to test this case (vmap over Tensor[1]), but the claim is
that it is because the boxed_reduction_batch_rule is agnostic to the shape
of the dimension being vmapped over. Previously it was not due to
the semantics of `squeeze`; this PR adds internal asserts to make it agnostic.
- there is a light test for vmap over the Tensor of shape [1] for
torch.sum as a sanity check.

### CODE REVIEW

File: aten/src/ATen/functorch/BatchRulesReduceOps.cpp [hunk 1]
```diff
@@ -368,58 +421,80 @@ std::tuple<Tensor,optional<int64_t>> searchsorted_batch_rule(
...
-  REDUCTION_BOXED(mean.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(mean.dim);
   m.impl("median", median_decomp);
-  REDUCTION_BOXED(median.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(median.dim);
   m.impl("min", min_decomp);
-  REDUCTION_BOXED(min.dim);
-  REDUCTION_BOXED(mode);
+  REDUCTION_WITH_KEEPDIM_ARG(min.dim);
+  REDUCTION_WITH_KEEPDIM_ARG(mode);
```
Comment (samdow) [inline aten/src/ATen/functorch/BatchRulesReduceOps.cpp:476]: (nothing needed for this PR) does this suggest that the original tests would have missed these as well? It might be good to have a follow up to extend the op infos for ones we need as a followup

**Review by samdow — COMMENTED** [review]
A little confused on why we're running the op instead of treating it like a no op but probably just not internalizing something

File: aten/src/ATen/functorch/BatchRulesReduceOps.cpp [hunk 2]
```diff
@@ -133,7 +156,39 @@ void boxed_reduction_batch_rule(const c10::OperatorHandle& op, torch::jit::Stack
...
   if (is_scalar_case) {
+    // NOTE: [boxed_reduction_batch_rule scalar tensor handling]
+    // Reduction operations in PyTorch have an edge case where they allow
+    // dim=0 and dim=-1 if the tensor has shape [].
+    //
+    // This can come up if we do something like
+    // vmap(lambda x: x.sum(0))(torch.tensor([10.])),
+    //
+    // In order to handle this edge case, we unsqueeze a dimension on the Tensor,
+    // run the operation (with dim=1 instead), and then process the output tensor.
```
[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesReduceOps.cpp:169]: What do you mean by "treating this like a no op"?

File: aten/src/ATen/functorch/BatchRulesReduceOps.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesReduceOps.cpp:476]: The original tests missed all of these cases. The edge case is vmap over a dimension of shape 1. The problem is not that the OpInfos are missing the test cases (they have them), the problem is that vmap testing generates dimensions of size 2 instead of size 1.

I didn't want to increase the testing time by adding vmap over dimension of shape 1 testing to all OpInfos. While we do not directly test this case, it is indirectly tested because the boxed_reduction_batch_rule is now written in a bdim-size-agnostic way.

File: aten/src/ATen/functorch/BatchRulesReduceOps.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesReduceOps.cpp:169]: Oh, you're wondering why we need to perform op to go from [B, 1] -> [B, 1], because it looks like a no-op.

It is still important that we run the op, because the op does something different depending on what the value inside the Tensor is.

Some cases:
- Let's say we have x = tensor(1.). Then x.sum(0) returns 1., which is indeed a no-op.
- However, consider x = tensor(float('nan')). x.sum(0) returns float('nan'), but x.nansum(0) returns 0.

**Review by samdow — APPROVED** [review]
Thanks for the explanations!
